### PR TITLE
Polyline tapping between points

### LINKED ISSUE #1: There is a Tappable Polygon ?
There is a Tappable Polygon ?  PolygonLayerOptions

### LINKED ISSUE #4: Detect taps between way points
Hi ! My use case is that I have many polylines close to each other, but not too many points on each polyline. How can I detect taps on the polyline with high precision without introducing new points and have to increase the radius of each point?

If I have to generate more points for this to work- any idea on how it will work with the order of 100.000-1.000.000 points in total? Will the performance still be good?

### PR DESCRIPTION
**Pull request by Mkohm**
This works pretty good :) Getting this to work was not as easy as initially thought! Need to test some more performance with more polylines. Feedback requested 👍

Close #4
Close #1

### CODE REVIEW

**Comment by tuarrep** [comment]
Thanks a lot for this PR! I will test it this weekend and make some feedbacks.

File: lib/flutter_map_tappable_polyline.dart [hunk 1]
```diff
@@ -99,28 +102,67 @@ class TappablePolylineLayer extends StatelessWidget {
...
-                        orElse: () => null) is Offset,
-                    orElse: () => null);
-
-                if (polyline is TaggedPolyline) onTap(polyline);
+                // Calculating taps in between points on the polyline. We
+                // iterate over all the segments in the polyline to find any
+                // matches with the tapped point within the
+                // pointerDistanceTolerance.
+                for (var i = 0; i < polylineOpts.polylines.length; i++) {
+                  Polyline currentPolyline = polylineOpts.polylines[i];
```
Comment (tuarrep) [inline lib/flutter_map_tappable_polyline.dart:110]: Please use `polylinesOpts.poylines.forEach`

File: lib/flutter_map_tappable_polyline.dart [hunk 2]
```diff
@@ -99,28 +102,67 @@ class TappablePolylineLayer extends StatelessWidget {
...
+                    // than the specified threshold, we have detected a tap
+                    // between two points.
+                    //
+                    // We start by calculating the length of all the sides using pythagoras.
+                    var a = distance(point1, point2);
+                    var b = distance(point1, tap);
+                    var c = distance(point2, tap);
+
+                    // To find the height when we only know the lengths of the sides, we can use Herons formula to get the Area.
+                    var semiPerimeter = (a + b + c) / 2.0;
```
Comment (tuarrep) [inline lib/flutter_map_tappable_polyline.dart:130]: Is `a - b - c <= polylineOpts.pointerDistanceTolenrance` sufficient?

File: lib/flutter_map_tappable_polyline.dart [hunk 3]
```diff
@@ -99,28 +102,67 @@ class TappablePolylineLayer extends StatelessWidget {
...
+                    var newTriangleBase =
+                        sqrt((hypotenus * hypotenus) - (height * height));
+                    var lengthDToOriginalSegment = newTriangleBase - a;
+
+                    if (height < polylineOpts.pointerDistanceTolerance &&
+                        lengthDToOriginalSegment <
+                            polylineOpts.pointerDistanceTolerance) {
+                      onTap(currentPolyline);
+                      return;
+                    } else {
```
Comment (tuarrep) [inline lib/flutter_map_tappable_polyline.dart:158]: Unnecessary as if returns

File: lib/flutter_map_tappable_polyline.dart [hunk 4]
```diff
@@ -99,28 +102,67 @@ class TappablePolylineLayer extends StatelessWidget {
...
+                      // We look in the next pair of points to see if we have a match there
+                      continue;
+                    }
+
+                    return;
```
Comment (tuarrep) [inline lib/flutter_map_tappable_polyline.dart:163]: Unneeded as previous lines return or continue

File: lib/flutter_map_tappable_polyline.dart [hunk 5]
```diff
@@ -1,8 +1,11 @@
 library flutter_map_tappable_polyline;
 
+import 'dart:math';
+
 import 'package:flutter/widgets.dart';
 import 'package:flutter_map/flutter_map.dart';
 import 'package:flutter_map/plugin_api.dart';
+import 'package:latlong/latlong.dart';
```
Comment (tuarrep) [inline lib/flutter_map_tappable_polyline.dart:8]: Unused

**Comment by HugoHeneault** [comment]
I even tested it with 2 different polylines and everything went well: taps event were sent to the right polyline. 👍
 
Simulator Screen Shot - iPhone SE - 2020-08-06 at 17 06 09

File: lib/flutter_map_tappable_polyline.dart (hunk 2 above)
Comment (Mkohm) [reply lib/flutter_map_tappable_polyline.dart:130]: I tested it using this formula, but there were some edge-cases that would not be covered properly.

Comment (tuarrep) [reply lib/flutter_map_tappable_polyline.dart:130]: OK. LGTM

**Comment by Mkohm** [comment]
Thanks for the feedback, I have updated the code accordingly!

File: lib/flutter_map_tappable_polyline.dart [hunk 6]
```diff
@@ -99,28 +101,60 @@ class TappablePolylineLayer extends StatelessWidget {
...
-                                    polylineOpts.pointerDistanceTolerance,
-                        orElse: () => null) is Offset,
-                    orElse: () => null);
-
-                if (polyline is TaggedPolyline) onTap(polyline);
+                // Calculating taps in between points on the polyline. We
+                // iterate over all the segments in the polyline to find any
+                // matches with the tapped point within t he
+                // pointerDistanceTolerance.
+                polylineOpts.polylines.forEach((element) {
```
Comment (tuarrep) [inline lib/flutter_map_tappable_polyline.dart:108]: Using `currentPolyline` here was great, could you please rename this variable to be clearer?

**Comment by tuarrep** [comment]
Thanks a lot @Mkohm for this great PR :pray:

**Pull request merged**